src/Microsoft.AspNet.OData.Shared/ODataAPIHandler.cs

```diff
@@ -64,129 +69,245 @@ public abstract class ODataAPIHandler<TStructuralType>: IODataAPIHandler where T
+                }
             }
         }
 
-        private object ApplyODataIdOnContainer(object obj, ODataPath odataPath)
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes", Justification = "We don't have a specific exception to catch.")]
+        internal static void ApplyHandlers(IODataAPIHandler odataApiHandler, IODataAPIHandler odataIdContainerHandler, object resource, IDictionary<string, object> keys, List<string> navigationProperties, out bool failedOperation)
         {
-            KeySegment keySegment = odataPath.LastOrDefault() as KeySegment;
-            Dictionary<string, object> keys = keySegment?.Keys.ToDictionary(x => x.Key, x => x.Value);
-            if (keys != null)
+            failedOperation = false;
+            DataModificationOperationKind operation = DataModificationOperationKind.Insert;
+
+            try
             {
-                TStructuralType returnedObject;
-                string error;
-                if (this.ParentObject.Equals(obj))
+                object[] handlerParams = new object[] { keys, null, null };
+
+                IODataAPIHandler handlerForGet = odataApiHandler;
+                ODataAPIResponseStatus getResponse = ODataAPIResponseStatus.NotFound;
+                object getObject = null;
+
+                if (odataIdContainerHandler != null)
                 {
-                    if (this.TryGet(keys, out returnedObject, out error) == ODataAPIResponseStatus.Success)
+                    handlerForGet = odataIdContainerHandler;
+                    getResponse = (ODataAPIResponseStatus)handlerForGet.GetType().GetMethod(nameof(TryGet)).Invoke(handlerForGet, handlerParams);
+                    getObject = handlerParams[1];
+                }
+
+                if (getResponse == ODataAPIResponseStatus.Success)
+                {
+                    operation = DataModificationOperationKind.Link;
+
+                    CopyProperties(getObject, resource, navigationProperties);
+
+                    object[] addRelatedObjectParams = new object[] { getObject, null };
+
+                    ODataAPIResponseStatus addRelatedObjectResponse = (ODataAPIResponseStatus)odataApiHandler.GetType().GetMethod(nameof(TryAddRelatedObject)).Invoke(odataApiHandler, addRelatedObjectParams);
+
+                    if (addRelatedObjectResponse == ODataAPIResponseStatus.Failure)
                     {
-                        return returnedObject;
+                        HandleFailedOperation(resource, operation, addRelatedObjectParams[1].ToString());
+                        failedOperation = true;
                     }
-                    else
+                }
+                else
+                {
+                    ODataAPIResponseStatus createObjectResponse = (ODataAPIResponseStatus)odataApiHandler.GetType().GetMethod(nameof(TryCreate)).Invoke(odataApiHandler, handlerParams);
+
```

Comment:
why are you not handling for a successful create operation like copy properties?

Reply:
Also you are not setting the `operation` in this create block

Reply:
Good catch. I inadvertently removed that code when doing some refactoring.
Fixed